hyperjumptech/monika
# Added: Templating feature for alert message configuration, and improvement in handling of alert in old format

# Monika Pull Request (PR)  

## What feature/issue does this PR add  
closes #350 and #351 

## How did you implement / how did you fix it  
1.  Use handlebars for templating
2. make improvement in conversion of  old alert format to new format
  - `status-not-2xx` is converted to: 
    ```json
    {
      "query": "response.status < 200 or response.status > 299",
      "subject": "",
      "message": "HTTP Status is {{ response.status }}, expecting 200"
    }
    ```
  - `response-time-greater-than-xx is converted to:
    ```json
    {
      "query": "response.time > X",
      "subject": "",
      "message": "Response time is {{ response.time }}ms, expecting less than Xms"
    }
    ```

These values are available for use inside the template:
- response.status: HTTP status code of the reponse
- response.time: the time it takes to perform a HTTP request
- response.size: size of the response in bytes
- response.headers: HTTP response headers
- response.body: HTTP response body (if content-type is JSON, it will be parsed automatically)

## How to test  
1. Run with config below:
    ```json
    {
      "notifications": [
        {
          "id": "desktop",
          "type": "desktop"
        }
      ],
      "probes": [
        {
          "id": "httpbin",
          "name": "probe httpbin",
          "requests": [
            {
              "method": "GET",
              "url": "https://reqres.in/api/users/3"
            }
          ],
          "alerts": [
            {
              "query": "response.body.data.id != 2",
              "subject": "Response data check",
              "message": "User id is {{ response.body.data.id }}, expecting 2"
            }
          ]
        }
      ]
    }
    ```
2. Play with the alert query to make the notification triggered
3. See the message on notification channels that you set in config
